contrib/aws/internal/sqs/sqs.go

```diff
@@ -0,0 +1,101 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package sqs
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sqs"
+	"github.com/aws/aws-sdk-go-v2/service/sqs/types"
+	"github.com/aws/smithy-go/middleware"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+const (
+	datadogKey           = "_datadog"
+	maxMessageAttributes = 10
+)
+
+type messageCarrier map[string]string
+
+func (carrier messageCarrier) Set(key, val string) {
+	carrier[key] = val
+}
+
+func EnrichOperation(ctx context.Context, in middleware.InitializeInput, operation string) {
+	switch operation {
+	case "SendMessage":
+		handleSendMessage(ctx, in)
+	case "SendMessageBatch":
+		handleSendMessageBatch(ctx, in)
+	}
+}
+
+func handleSendMessage(ctx context.Context, in middleware.InitializeInput) {
+	params, ok := in.Parameters.(*sqs.SendMessageInput)
+	if !ok {
+		log.Debug("Unable to read SendMessage params")
+		return
+	}
+
+	if params.MessageAttributes == nil {
+		params.MessageAttributes = make(map[string]types.MessageAttributeValue)
+	}
+
+	injectTraceContext(ctx, params.MessageAttributes)
+}
+
+func handleSendMessageBatch(ctx context.Context, in middleware.InitializeInput) {
+	params, ok := in.Parameters.(*sqs.SendMessageBatchInput)
+	if !ok {
+		log.Debug("Unable to read SendMessageBatch params")
+		return
+	}
+
+	for i := range params.Entries {
+		entryPtr := &params.Entries[i]
+		if entryPtr.MessageAttributes == nil {
+			entryPtr.MessageAttributes = make(map[string]types.MessageAttributeValue)
+		}
```

Comment:
Can you help me understand why its necessary to grab the pointer here? Why not just

```go
for _, entry := range params.Entries {
  if entry.MessageAttributes == nil {
    entry.MessageAttributes = make(map[string]types.MessageAttributeValue)
  }
  injectTraceContext(ctx, entry.MessageAttributes)
}
```

Reply:
We can't use `for _, entry` because we modify a copy of the entry instead of the original entry. Then the unit tests and manual tests will fail. 

However, we don't actually need a pointer. I just tested it and as long as we index the array each time, it works fine.

```
for i := range params.Entries {
  if params.Entries[i].MessageAttributes == nil {
    params.Entries[i].MessageAttributes = make(map[string]types.MessageAttributeValue)
  }
  injectTraceContext(ctx, params.Entries[i].MessageAttributes)
}
```